lib/common/socket.c

```diff
@@ -1470,3 +1470,116 @@ void h2o_sliding_counter_stop(h2o_sliding_counter_t *counter, uint64_t now)
     /* recalc average */
     counter->average = counter->prev.sum / (sizeof(counter->prev.slots) / sizeof(counter->prev.slots[0]));
 }
+
+#if H2O_USE_DTRACE && defined(__linux__)
+#include <linux/bpf.h>
+#include <linux/unistd.h>
+#include "h2o-probes.h"
+#include "include/h2o/ebpf.h"
+#include <sys/stat.h>
+
+static __thread int tracing_map_fd = -1;
+static __thread uint64_t tracing_map_last_attempt = 0;
+
+static void open_tracing_map(h2o_socket_t *sock)
+{
+    // only check every second
+    uint64_t now = h2o_now(h2o_socket_get_loop(sock));
+    if (tracing_map_last_attempt - now < 1000)
+        return;
+
+    tracing_map_last_attempt = now;
+
+    // check if map exists at path
+    struct stat s;
+    if (stat(&H2O_EBPF_MAP_PATH[0], &s) == -1) {
+        // map path unavailable, cleanup fd if needed and leave
+        if (tracing_map_fd > 0) {
+            close(tracing_map_fd);
+            tracing_map_fd = -1;
+        }
+        return;
+    }
+
+    if (tracing_map_fd > 0)
+        return; // map still exists and we have a fd
+
+    // map exists, try connect
+    union bpf_attr attr;
+    memset(&attr, 0, sizeof(attr));
+    attr.pathname = (uint64_t)&H2O_EBPF_MAP_PATH[0];
+    tracing_map_fd = syscall(__NR_bpf, BPF_OBJ_GET, &attr, sizeof(attr));
+}
+
+int lookup_map(const void *key, const void *value)
+{
+    union bpf_attr attr;
+    memset(&attr, 0, sizeof(attr));
+    attr.map_fd = tracing_map_fd;
+    attr.key = (uint64_t)key;
+    attr.value = (uint64_t)value;
+    return syscall(__NR_bpf, BPF_MAP_LOOKUP_ELEM, &attr, sizeof(attr)) == -1 ? -1 : 1; // return 1 if found, -1 otherwise
+}
+
+inline void set_ebpf_map_key_tuples(struct sockaddr *sa, uint8_t *ip, uint16_t *port)
+{
+    if (sa->sa_family == AF_INET) {
+        struct sockaddr_in *sin = (void *)sa;
+        memcpy(ip, &sin->sin_addr, sizeof(sin->sin_addr));
+        *port = sin->sin_port;
+    } else if (sa->sa_family == AF_INET6) {
+        struct sockaddr_in6 *sin = (void *)sa;
+        memcpy(ip, &sin->sin6_addr, sizeof(sin->sin6_addr));
+        *port = sin->sin6_port;
+    }
+}
+
+inline int init_ebpf_map_key(h2o_ebpf_map_key_t *key, h2o_socket_t *sock)
+{
+    struct sockaddr_storage sockname, peername;
+    unsigned int sock_type, sock_type_len = sizeof(sock_type_len);
+    memset(key, 0, sizeof(h2o_ebpf_map_key_t));
```

Comment:
I think it's our convention to use `memset(key, 0, sizeof(*key))` (notice the argument to sizeof), so that the correspondence between the first argument and its size can be managed by the compiler.